sklearn/mixture/_base.py

```diff
@@ -252,28 +256,35 @@ def fit_predict(self, X, y=None):
 
             lower_bound = -np.inf if do_init else self.lower_bound_
 
-            for n_iter in range(1, self.max_iter + 1):
-                prev_lower_bound = lower_bound
+            if self.max_iter == 0:
+                best_params = self._get_parameters()
+                best_n_iter = 1
+            else:
+                for n_iter in range(1, self.max_iter + 1):
+                    prev_lower_bound = lower_bound
 
-                log_prob_norm, log_resp = self._e_step(X)
-                self._m_step(X, log_resp)
-                lower_bound = self._compute_lower_bound(log_resp, log_prob_norm)
+                    log_prob_norm, log_resp = self._e_step(X)
+                    self._m_step(X, log_resp)
+                    lower_bound = self._compute_lower_bound(log_resp, log_prob_norm)
 
-                change = lower_bound - prev_lower_bound
-                self._print_verbose_msg_iter_end(n_iter, change)
+                    change = lower_bound - prev_lower_bound
+                    self._print_verbose_msg_iter_end(n_iter, change)
 
-                if abs(change) < self.tol:
-                    self.converged_ = True
-                    break
+                    if abs(change) < self.tol:
+                        self.converged_ = True
+                        break
 
-            self._print_verbose_msg_init_end(lower_bound)
+                self._print_verbose_msg_init_end(lower_bound)
 
-            if lower_bound > max_lower_bound or max_lower_bound == -np.inf:
-                max_lower_bound = lower_bound
-                best_params = self._get_parameters()
-                best_n_iter = n_iter
+                if lower_bound > max_lower_bound or max_lower_bound == -np.inf:
+                    max_lower_bound = lower_bound
+                    best_params = self._get_parameters()
+                    best_n_iter = n_iter
 
-        if not self.converged_:
+        # Should only warn about convergence if max_iter > 0, otherwise
+        # the user is assumed to have used 0-iters initialization
+        # to get the initial means.
+        if not self.converged_ and self.max_iter > 0:
```

Comment:
you could just put this block into the `else` block so that it can only be triggered if max_iter > 0

Reply:
Hey, thanks for reviewing our PR!

The benefit of doing this doesn't seem as clear to me. I understand that doing a for-else will result in the same behavior but it seems to me that, in its current state, the warning will only get triggered if max_iter > 0 anyway. Could you clarify? I might have misunderstood your suggestion, too.

Kindly,

Alberto

Reply:
sorry I missed 1 level of indentation. It's fine as is indeed.